services/runservice/types/runevent.go

```diff
@@ -1,18 +1,115 @@
 package types
 
 import (
+	"encoding/json"
+	"time"
+
+	"github.com/sorintlab/errors"
+
 	"agola.io/agola/internal/sqlg"
 	"agola.io/agola/internal/sqlg/sql"
 )
 
+type RunEventType string
+
+const (
+	RunPhaseChanged RunEventType = "run_phase_changed"
+
+	RunEventDataVersion = 1
+)
+
 type RunEvent struct {
 	sqlg.ObjectMeta
 
 	Sequence uint64
 
-	RunID  string
-	Phase  RunPhase
-	Result RunResult
+	RunEventType RunEventType
+	RunID        string
+	Phase        RunPhase
+	Result       RunResult
+
+	Annotations map[string]string
+
+	Data        any
+	DataVersion uint64
+}
+
+func (e *RunEvent) PreJSON() error {
+	switch e.DataVersion {
+	case 1:
+		e.Data = &RunEventData{}
+	default:
+		return errors.Errorf("unknown runevent data version: %d", e.DataVersion)
+	}
+
+	return nil
+}
+
+func (e *RunEvent) UnmarshalJSON(data []byte) error {
+	type runEvent RunEvent
+
+	t := (*runEvent)(e)
+
+	if err := json.Unmarshal(data, &t); err != nil {
+		return errors.WithStack(err)
+	}
+	d := e.Data
+
+	if err := e.PreJSON(); err != nil {
+		return errors.Wrap(err, "prejson error")
+	}
+
+	inDataJSON, err := json.Marshal(d)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	if err := json.Unmarshal(inDataJSON, e.Data); err != nil {
+		return errors.WithStack(err)
+	}
+
+	return nil
+}
```

Comment:
I can't understand how this could work. This is what I'll use:

```
func (e *RunEvent) UnmarshalJSON(data []byte) error {
	type origRunEvent RunEvent
	type runEvent struct {
		*origRunEvent
		Data json.RawMessage
	}

	v := &runEvent{}

	if err := json.Unmarshal(data, &v); err != nil {
		return err
	}

	*e = RunEvent(*v.origRunEvent)

	switch v.DataVersion {
	case 1:
		e.Data = &RunEventData{}
	default:
		return errors.Errorf("unknown runevent data version: %d", e.DataVersion)
	}

	if err := json.Unmarshal(v.Data, &e.Data); err != nil {
		return err
	}

	return nil
}
```